src/v/storage/offset_translator.cc

```diff
@@ -253,65 +254,48 @@ ss::future<> offset_translator::truncate(model::offset offset) {
     co_await _checkpoint_lock.with([this] { return do_checkpoint(); });
 }
 
-ss::future<> offset_translator::prefix_truncate(model::offset offset) {
+ss::future<> offset_translator::prefix_truncate(
+  model::offset offset,
+  std::optional<model::offset_delta> force_truncate_delta) {
     if (_filtered_types.empty()) {
         co_return;
     }
 
-    if (offset > _highest_known_offset) {
-        throw std::runtime_error{_logger.format(
-          "trying to prefix truncate offset translator at offset {} which "
-          "is > highest_known_offset {}",
-          offset,
-          _highest_known_offset)};
-    }
-
-    if (!_state->prefix_truncate(offset)) {
+    if (unlikely(offset > _highest_known_offset)) {
+        if (!force_truncate_delta.has_value()) {
+            throw std::runtime_error{_logger.format(
+              "trying to prefix truncate offset translator at offset {} which "
+              "is > highest_known_offset {}",
+              offset,
+              _highest_known_offset)};
+        }
+        // If the truncation is for past the end of the log (e.g. during
+        // bootstrap, and a bug has caused us to desire a new start offset past
+        // the end of the log, allow it.
+        *_state = storage::offset_translator_state(
+          _state->ntp(), offset, force_truncate_delta.value());
+        _highest_known_offset = offset;
+    } else if (!_state->prefix_truncate(offset)) {
         co_return;
     }
 
     ++_map_version;
 
-    vlog(
-      _logger.debug,
-      "prefix_truncate at offset: {}, new state: {}",
+    vlogl(
+      _logger,
+      force_truncate_delta.has_value() ? ss::log_level::info
+                                       : ss::log_level::debug,
+      "prefix_truncate at offset: {}, force_truncate_delta: {}, new state: {}",
       offset,
+      force_truncate_delta,
       _state);
 
     co_await _checkpoint_lock.with([this] { return do_checkpoint(); });
 }
 
 ss::future<>
 offset_translator::prefix_truncate_reset(model::offset offset, int64_t delta) {
-    if (_filtered_types.empty()) {
```

Comment:
just a little confused by the commit message. does this commit change any behavior, or generalize `prefix_truncate` to handle what had been done by both `prefix_truncate_reset` and `prefix_truncate`?

Reply:
Sorry it wasn't clear: it's just generalizing `prefix_truncate`, exactly